core/common/src/test/java/alluxio/heartbeat/SleepingTimerForCronExpressionIntervalSupplierTest.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.heartbeat;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+
+import alluxio.Constants;
+import alluxio.clock.ManualClock;
+import alluxio.time.Sleeper;
+
+import org.apache.logging.log4j.core.util.CronExpression;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.slf4j.Logger;
+
+import java.text.DateFormat;
+import java.text.ParseException;
+import java.text.SimpleDateFormat;
+import java.time.Duration;
+import java.util.Date;
+
+/**
+ * Unit tests for {@link SleepingTimer}.
+ */
+public final class SleepingTimerForCronExpressionIntervalSupplierTest {
+  private static final String THREAD_NAME = "cron-test-thread-name";
+  private static final long INTERVAL_MS = 10 * Constants.MINUTE_MS;
+  private Logger mMockLogger;
+  private ManualClock mFakeClock;
+  private Sleeper mMockSleeper;
+  private long mAllSleepTimeMs;
+
+  @Before
+  public void before() throws InterruptedException {
+    mMockLogger = mock(Logger.class);
+    mFakeClock = new ManualClock();
+    mMockSleeper = mock(Sleeper.class);
+    doAnswer((invocation) -> {
+      Duration duration = invocation.getArgument(0);
+      mFakeClock.addTime(duration);
+      mAllSleepTimeMs += duration.toMillis();
+      return null;
+    }).when(mMockSleeper).sleep(any(Duration.class));
+  }
+
+  /**
+   * Tests that the cron timer will attempt to run at the same interval, independently of how
+   * long the execution between ticks takes. For example, if the interval is 100ms and execution
+   * takes 80ms, the timer should sleep for only 20ms to maintain the regular interval of 100ms.
+   */
+  @Test
+  public void maintainInterval() throws Exception {
+    SleepingTimer timer =
+        new SleepingTimer(THREAD_NAME, mMockLogger, mFakeClock, mMockSleeper,
+            () -> {
+              try {
+                return new CronExpressionIntervalSupplier(
+                    new CronExpression("* 30-59 0-1,4-9,13-23 * * ? *"), INTERVAL_MS);
```

Comment:
Looks like this is not the standard [crontab](https://en.wikipedia.org/wiki/Cron#Overview) expression. When we add the docs, a link to the cron syntax specification should be included. https://logging.apache.org/log4j/2.x/log4j-core/apidocs/org/apache/logging/log4j/core/util/CronExpression.html

Reply:
Is there a reason we use log4j's implementation of cron? Is this cron syntax commonly used by log4j or other utilities?

Reply:
@dbw9580 Log4j cron expression is used for config `RollingFileAppender` and `TimeBasedRollingPolicy`, it can control the rolling of log file, something like 

```properties
appender.rolling.type = RollingFile
appender.rolling.name = RollingFile
appender.rolling.fileName = logs/app.log
appender.rolling.filePattern = logs/app-%d{yyyy-MM-dd}.log
appender.rolling.layout.type = PatternLayout
appender.rolling.layout.pattern = %d{yyyy-MM-dd HH:mm:ss} [%t] %-5p %c{1}:%L - %m%n
appender.rolling.policies.type = Policies
appender.rolling.policies.time.type = TimeBasedTriggeringPolicy
appender.rolling.policies.time.interval = 1
appender.rolling.policies.time.modulate = true
appender.rolling.policies.time.cron = 0 0 1 * * ?
```